### ISSUE TITLE
fastCopyAndTranspose segfaults (migrated from Trac #766)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#1373]
Original ticket http://projects.scipy.org/numpy/ticket/766
Reported 2008-04-30 by trac user dave.hirschfeld, assigned to unknown.

The below code causes fastCopyAndTranspose to segfault.

> It looks like the segfault occurs in the printing of the result.    If 
> you set y = fastCopyAndTranspose(x)  you don't get a segfault until you 
> try to print it (actually **repr** it). 
> 
> -Travis

from datetime import datetime, timedelta

from numpy import asarray, fastCopyAndTranspose

timestamps = [datetime(2007,1,1)]

while timestamps[-1] < datetime(2007,12,31):
    timestamps.append(timestamps[-1] + timedelta(1))

x = [asarray(timestamps),asarray(range(len(timestamps)))]

fastCopyAndTranspose(x)

**Comment by thouis** [comment]
Comment in Trac by atmention:alanmcintyre, 2008-05-14

I poked around with the test case under the debugger for a while, and here's some observations. Maybe once I get familiar enough with all the details it'll make sense to me, but right now I'm not able to spot where the problem actually occurs.  If anybody can fill in any details or suggest something in particular to look at, I'll be happy to do so.

Here's a case that doesn't segfault:

```
from datetime import datetime
from numpy import fastCopyAndTranspose,array
x= array([[datetime(2007,1,1)]])
y = fastCopyAndTranspose(x)
print repr(y)
```

But adding 'del x' before taking repr(y) causes a segfault (or sometimes an illegal instruction):

```
from datetime import datetime
from numpy import fastCopyAndTranspose,array
x= array([[datetime(2007,1,1)]])
y = fastCopyAndTranspose(x)
del x
print repr(y)
```

It appears that the segfault occurs when attempting to call v->ob_type->tp_str in _PyObject_Str.

The following example:
    from numpy import fastCopyAndTranspose

```
class A: pass
y=fastCopyAndTranspose([[A()]])
print repr(y)
```

prints the following output:
    array([['numpy.ndarray' object has no attribute '_format']], dtype=object)
This looks a little like an exception being swallowed up somewhere, but I'm not sure.  Attempting to do something else with A after calling repr on the array:
    from numpy import fastCopyAndTranspose

```
class A: pass
y=fastCopyAndTranspose([[A()]])
print repr(y)
x=A()
```

causes a "Fatal Python error: deletion of interned string failed" message when the interpreter exits.

**Comment by thouis** [comment]
Comment in Trac by atmention:jarrodmillman, 2008-05-15

**Comment by thouis** [comment]
Comment in Trac by atmention:cournape, 2008-11-20

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2011-03-04

Seems like a refcounting error.

Indeed, `PyArray_CopyAndTranspose` does not handle refcounting the items at all --- this should be fixed.

**Comment by thouis** [comment]
Comment in Trac by atmention:mwiebe, 2011-03-23

https://github.com/numpy/numpy/pull/60 is a fix for it.

**Comment by thouis** [comment]
Comment in Trac by atmention:mwiebe, 2011-03-25

Fixed in 73ea5e7be76d77b6ac67 (master), 196186daeb5b25a93d9e (1.6.x).
